### PR TITLE
fixes greedy dots in extension parsing

### LINKED ISSUE #9765: RequestHandler json dots in pass parameter
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.3.8
* Platform and Target: MacOS 10.11.6, PHP 5.6.14

### What you did
I have a route scope api running with extension json, like:

```
Router::scope('/api', function ($routes) {
    $routes->extensions(['json']);
    $routes->resources('Applications', [
            'actions' => ['view'],
            'pass' => ['id'],
            'id' => '[.a-zA-Z0-9_-]+'
        ]
    ); 
});

```

I have routes like _localhost/myapp/api/applications/com.mycompany.myapp.json_, which were working fine for nearly 2 years now (I'm not quite sure from version I updated, but I guess it has been in the 3.3. branch). 
It extracted _com.mycompany.myapp_ as $id, which i could you use in ApplicationsController::view(). 

 
### What happened
But today I recognized, that I recieve _com.mycompany.myapp.json_ . It seems like the RequestHandler won't cut away the .json anymore, when there are . (dots) in the pass part of the URI.
When i.e. I'm using _localhost/myapp/api/applications/wurstwasser.json_ I recieve wurstwasser (without trailing json).  

### What you expected to happen
I expect to get _com.mycompany.myapp_ from _localhost/myapp/api/applications/com.mycompany.myapp.json_

### PR DESCRIPTION
**Pull request by thinkingmedia**
closes #9765 

- removes the usage of regex in parsing extensions, and uses a simple "string ends with" check for extensions.
- updates `extensions()` with `setExtensions()` and `getExtensions()`
- added unit tests for changes
- adds stub class to expose protected methods (unit test only)

### CODE REVIEW

File: src/Routing/Route/Route.php [hunk 1]
```diff
@@ -346,20 +368,12 @@ public function parse($url, $method = '')
...
-            return [$url, null];
-        }
-        $ext = strtolower($match[1]);
-        $len = strlen($match[1]);
-        foreach ($this->_extensions as $name) {
-            if (strtolower($name) === $ext) {
-                $url = substr($url, 0, ($len + 1) * -1);
-
-                return [$url, $ext];
+        if (count($this->_extensions) && strpos($url, '.')) {
```
Comment (markstory) [inline src/Routing/Route/Route.php]: I think you want `strpos(...) !== false` here, as the `.` could be in the 0th spot.

File: src/Routing/Route/Route.php [hunk 2]
```diff
@@ -346,20 +368,12 @@ public function parse($url, $method = '')
...
+            foreach ($this->_extensions as $ext) {
+                $len = strlen($ext) + 1;
+                if (substr($url, -$len) === '.' . $ext) {
+                    return [substr($url, 0, $len * -1), $ext];
+                }
```
Comment (markstory) [inline src/Routing/Route/Route.php]: Would building a regexp and applying it be simpler here?

File: tests/TestCase/Routing/Route/RouteTest.php [hunk 3]
```diff
@@ -20,6 +20,29 @@
...
+ * Used to expose protected methods for testing.
+ */
+class RouteProtected extends Route
+{
+    /**
+     * @return array
+     */
+    public function peekExtensions()
+    {
+        return $this->_extensions;
```
Comment (markstory) [inline tests/TestCase/Routing/Route/RouteTest.php]: How is this different than `getExtensions()`?

Comment (thinkingmedia) [inline tests/TestCase/Routing/Route/RouteTest.php]: I just wanted to test the setter function without being dependent upon the getter function. I can use the getter if you want, but I'm just being defensive with the tests.

File: src/Routing/Route/Route.php (hunk 1 above)
Comment (thinkingmedia) [inline src/Routing/Route/Route.php]: good catch, thanks.

File: tests/TestCase/Routing/Route/RouteTest.php (hunk 3 above)
Comment (markstory) [inline tests/TestCase/Routing/Route/RouteTest.php]: But you have another getter :smile:

File: src/Routing/Route/Route.php (hunk 2 above)
Comment (markstory) [inline src/Routing/Route/Route.php]: I was thinking the regex would be simpler code. I'm not sure what the performance difference would be.

Comment (thinkingmedia) [inline src/Routing/Route/Route.php]: - `$lan = strlen($ext) + 1` is from the original code.
- `substr($url, -$len)` should not be confusing to understand.
- the last `substr` is from the original ode.

Comment (markstory) [inline src/Routing/Route/Route.php]: Yeah, I'm just being dense.

**Pull request merged**